packages/sites/src/use-sites-list-sorting.tsx

```diff
@@ -153,6 +157,74 @@ function sortSitesByLastInteractedWith< T extends SiteDetailsForSorting >(
 	return sortOrder === 'desc' ? sortedItems : sortedItems.reverse();
 }
 
+export function sortSitesByStaging< T extends SiteDetailsForSorting >( sites: T[] ) {
+	type SitesByIdType = {
+		index: number;
+		site: T;
+	};
+	const sitesById = sites.reduce< Record< number, SitesByIdType > >(
+		( acc, site, currentIndex ) => {
+			acc[ site.ID ] = {
+				index: currentIndex,
+				site,
+			};
+			return acc;
+		},
+		{}
+	);
+
+	const visited = {} as Record< number, boolean >;
+	const sortedItems = sites.reduce< T[] >( ( acc, site ) => {
+		// We have already visited this site, therefore, there is no need to proceed any further.
+		if ( visited[ site.ID ] ) {
+			return acc;
+		}
+		// Site is staging but we haven't visit its production site yet...
+		// The production site exists in the site map, and
+		// that means that we are going to visit it later on.
+		// Don't add it to the list yet.
+		if (
+			site?.is_wpcom_staging_site &&
+			site.options?.wpcom_production_blog_id &&
+			! visited[ site.options?.wpcom_production_blog_id ] &&
+			sitesById[ site.options?.wpcom_production_blog_id ]
+		) {
+			return acc;
+		}
+
+		// If the production site associated with this staging site is filtered out,
+		// or if we have a production site, add it to the list.
+		acc.push( site );
+		visited[ site.ID ] = true;
+
+		// Sorting is useful when dealing with multiple staging sites.
+		// The sites are already sorted by 'sortSitesByLastInteractedWith',
+		// we want to maintain that order in case some of the staging sites are
+		// filtered out.
```

Comment:
Oy vey!

Reply:
I didn't know what Oy vey meant. Does it really use to express dismay, frustration, or grief?

Reply:
@kozer https://en.wikipedia.org/wiki/Oy_vey 😊

Reply:
(it's a general reaction to the idea that we had to write some complex code in order to achieve something seemingly simple)